### ISSUE TITLE
Assembly errors with clang-10

### ISSUE THREAD

**Issue by yurivict** [sadko4u/lsp-plugins#154]
Log: https://people.freebsd.org/~yuri/lsp-plugins-lv2-1.1.26_1.log

errors:
```
/wrkdirs/usr/ports/audio/lsp-plugins-lv2/work/lsp-plugins-lsp-plugins-1.1.26/include/dsp/arch/x86/avx/filters/transfer.h:42:13: error: expected relocatable expression
            __ASM_EMIT("vbroadcastss        0x18(%[c]), %%ymm5")                // y5   = b2
            ^
/wrkdirs/usr/ports/audio/lsp-plugins-lv2/work/lsp-plugins-lsp-plugins-1.1.26/include/common/types.h:263:57: note: expanded from macro '__ASM_EMIT'
#define __ASM_EMIT(code)                        code "\n\t"
                                                        ^
<inline asm>:7:2: note: instantiated into assembly here
        vmovaps             %ymm0, 0x00 + (%rsp)
        ^
...etc...
```

**Comment by sadko4u** [comment]
For me it just crashes:
```
gmake

...

bits.cpp:48:20: warning: unused variable '__rb_masks' [-Wunused-const-variable]
    const uint32_t __rb_masks[] =
                   ^
1 warning generated.
  c++ native.cpp
  c++ x86.cpp
  c++ sse.cpp
fatal error: error in backend: Cannot select: 0x80666ad68: v4f32 = fp_extend 0x80666add0
  0x80666add0: f32,ch = CopyFromReg 0x8053690e8, Register:f32 %1
    0x80666ad00: f32 = Register %1
In function: _ZN3sse4fillEPffm
Stack dump:
0.      Program arguments: c++ -o /usr/home/sadko/git/lsp-plugins/.build/src/dsp/sse.o -c sse.cpp -fPIC -m64 -std=c++98 -fno-exceptions -fno-rtti -fdata-sections -ffunction-sections -fno-asynchronous-unwind-tables -fvisibility=hidden -pipe -Wall -DLSP_MAIN_VERSION="1.1.26" -DLSP_INSTALL_PREFIX="/usr/local" -O2 -DLSP_NO_EXPERIMENTAL -mmmx -m3dnow -msse -I/usr/home/sadko/git/lsp-plugins/include 
1.      <eof> parser at end of file
2.      Code generation
3.      Running pass 'Function Pass Manager' on module 'sse.cpp'.
4.      Running pass 'X86 DAG->DAG Instruction Selection' on function '@_ZN3sse4fillEPffm'
 #0 0x00000000038320be (/usr/bin/c+++0x38320be)
 #1 0x0000000003830335 (/usr/bin/c+++0x3830335)
 #2 0x0000000003834198 (/usr/bin/c+++0x3834198)
 #3 0x000000000383413e (/usr/bin/c+++0x383413e)
 #4 0x00000000038334f3 (/usr/bin/c+++0x38334f3)
 #5 0x00000000016dc7b5 (/usr/bin/c+++0x16dc7b5)
 #6 0x00000000046ec4a6 (/usr/bin/c+++0x46ec4a6)
 #7 0x00000000046ec573 (/usr/bin/c+++0x46ec573)
 #8 0x00000000040d206f (/usr/bin/c+++0x40d206f)
 #9 0x00000000040d18d2 (/usr/bin/c+++0x40d18d2)
#10 0x0000000002f1a46e (/usr/bin/c+++0x2f1a46e)
#11 0x00000000040ca9cc (/usr/bin/c+++0x40ca9cc)
#12 0x00000000040ca099 (/usr/bin/c+++0x40ca099)
#13 0x00000000040c97f9 (/usr/bin/c+++0x40c97f9)
#14 0x00000000040c7122 (/usr/bin/c+++0x40c7122)
#15 0x0000000002f1104b (/usr/bin/c+++0x2f1104b)
#16 0x000000000445771e (/usr/bin/c+++0x445771e)
#17 0x0000000003dc0b81 (/usr/bin/c+++0x3dc0b81)
#18 0x0000000003dc0e33 (/usr/bin/c+++0x3dc0e33)
#19 0x0000000003dc130a (/usr/bin/c+++0x3dc130a)
#20 0x000000000172a278 (/usr/bin/c+++0x172a278)
#21 0x0000000001724076 (/usr/bin/c+++0x1724076)
#22 0x00000000022c66c3 (/usr/bin/c+++0x22c66c3)
#23 0x0000000001c394d6 (/usr/bin/c+++0x1c394d6)
#24 0x0000000002ccd6c9 (/usr/bin/c+++0x2ccd6c9)
#25 0x00000000016eb076 (/usr/bin/c+++0x16eb076)
#26 0x00000000016dc394 (/usr/bin/c+++0x16dc394)
#27 0x00000000016e8b74 (/usr/bin/c+++0x16e8b74)
#28 0x0000000002e99577 (/usr/bin/c+++0x2e99577)
#29 0x0000000003834122 (/usr/bin/c+++0x3834122)
#30 0x0000000002e98d6e (/usr/bin/c+++0x2e98d6e)
#31 0x0000000002dae404 (/usr/bin/c+++0x2dae404)
#32 0x0000000002dae94c (/usr/bin/c+++0x2dae94c)
#33 0x0000000002d72d3c (/usr/bin/c+++0x2d72d3c)
#34 0x00000000016e8695 (/usr/bin/c+++0x16e8695)
c++: error: clang frontend command failed with exit code 70 (use -v to see invocation)
FreeBSD clang version 10.0.1 (git@github.com:llvm/llvm-project.git llvmorg-10.0.1-0-gef32c611aa2)
Target: x86_64-unknown-freebsd12.2
Thread model: posix
InstalledDir: /usr/bin
c++: note: diagnostic msg: PLEASE submit a bug report to https://bugs.freebsd.org/submit/ and include the crash backtrace, preprocessed source, and associated run script.
c++: note: diagnostic msg: 
********************

PLEASE ATTACH THE FOLLOWING FILES TO THE BUG REPORT:
Preprocessed source(s) and associated run script(s) are located at:
c++: note: diagnostic msg: /tmp/sse-99e394.cpp
c++: note: diagnostic msg: /tmp/sse-99e394.sh
c++: note: diagnostic msg: 

********************
gmake[2]: *** [Makefile:70: /usr/home/sadko/git/lsp-plugins/.build/src/dsp/sse.o] Error 70
gmake[1]: *** [Makefile:71: dsp] Error 2
gmake: *** [Makefile:175: compile] Error 2
```

The version of clang:
```
[sadko@freebsd ~/git/lsp-plugins]$ clang --version
FreeBSD clang version 10.0.1 (git@github.com:llvm/llvm-project.git llvmorg-10.0.1-0-gef32c611aa2)
Target: x86_64-unknown-freebsd12.2
Thread model: posix
InstalledDir: /usr/bin
```

**Comment by yurivict** [comment]
> For me it just crashes:

This has been fixed in their master, you need to try their development version.

**Comment by sadko4u** [comment]
> > For me it just crashes:
> 
> This has been fixed in their master, you need to try their development version.

How can I install it from pkg?

**Comment by yurivict** [comment]
You can install the latest FreeBSD 12.2 - it has the patched clang-10.

**Comment by sadko4u** [comment]
> You can install the latest FreeBSD 12.2 - it has the patched clang-10.

```
[sadko@freebsd ~/git/lsp-dsp-lib]$ uname -a
FreeBSD freebsd 12.2-RELEASE FreeBSD 12.2-RELEASE r366954 GENERIC  amd64
[sadko@freebsd ~/git/lsp-dsp-lib]$ clang --version
FreeBSD clang version 10.0.1 (git@github.com:llvm/llvm-project.git llvmorg-10.0.1-0-gef32c611aa2)
Target: x86_64-unknown-freebsd12.2
Thread model: posix
InstalledDir: /usr/bin
```
It seems that I have the same compiler as mentioned in build log. It crashes for me.

**Comment by yurivict** [comment]
You use ```12.2-RELEASE```. RELEASE can't be updated and still has that bug in the compiler.

You need ```12.2-STABLE```. ISO files are here: https://download.freebsd.org/ftp/snapshots/amd64/amd64/ISO-IMAGES/12.2/ The latest one is https://download.freebsd.org/ftp/snapshots/amd64/amd64/ISO-IMAGES/12.2/FreeBSD-12.2-STABLE-amd64-20201112-r367603-disc1.iso

In ```12.2-STABLE```:
1. ```cd /usr/ports/audio/lsp-plugins-lv2```
2. comment out the ```USE_GCC=``` line
3. ```sudo pkg install -A $(make missing)```
4. ```make```

**Comment by sadko4u** [comment]
Uuggh. Just installed 12.2-RELEASE this night.
Can I update it to 12.2-STABLE with freebsd-update?

**Comment by yurivict** [comment]
```freebsd-update -r 12.2-STABLE upgrade``` should work.

**Comment by sadko4u** [comment]
Nope, seems to be not working.
```
Fetching metadata signature for 12.2-STABLE from update1.freebsd.org... failed.
Fetching metadata signature for 12.2-STABLE from update4.freebsd.org... failed.
Fetching metadata signature for 12.2-STABLE from update2.freebsd.org... failed.
No mirrors remaining, giving up.

This may be because upgrading from this platform (amd64)
or release (12.2-STABLE) is unsupported by freebsd-update. Only
platforms with Tier 1 support can be upgraded by freebsd-update.
See https://www.freebsd.org/platforms/index.html for more info.

If unsupported, FreeBSD must be upgraded by source.
```

OK, let's do clear FreeBSD installation.

**Comment by sadko4u** [comment]
I think clang has still bugs related to the AVX support.
Here's the description of the 'o' assembly constraint:
>"m" : A memory operand is allowed, with any kind of address that the machine supports in general.
>"o" : A memory operand is allowed, but only if the address is offsettable. ie, **adding a small offset to the address gives a valid address.**

So, such trick works with SSE:
```
                __ASM_EMIT("movaps      0x00 + %[CC], %%xmm5")      /* xmm5  = 1 */
                __ASM_EMIT("xorps       0x10 + %[CC], %%xmm1")      /* xmm1  = -i0 */
```

BUt doesn't work with AVX:
```
                __ASM_EMIT("vmovaps             %%ymm0, 0x00 + %[fp]")              // x0   = t0
                __ASM_EMIT("vmovaps             %%ymm1, 0x20 + %[fp]")              // x1   = t1
                __ASM_EMIT("vmovaps             %%ymm2, 0x40 + %[fp]")              // x2   = t2
```

**Comment by yurivict** [comment]
Are you able to create the bug report with a minimal testcase?

https://bugs.llvm.org/

**Comment by sadko4u** [comment]
Here is the test case.

**test_sse.cpp**:
```
#include <stdint.h>

static const uint32_t XS[4] = { 1, 2, 3, 4 };
int32_t XD[4] = { 0, 0, 0, 0 };

void test_avx()
{
    __asm__ __volatile__(
        "movups 0x00 + %[xs], %%xmm0\n\t"
        "movups %%xmm0, 0x00 + %[xd]\n\t"
        :
        : [xs] "o" (XS),
          [xd] "o" (XD)
        : "%xmm0"
    );
}
```


**test_avx.cpp**:
```
#include <stdint.h>

static const uint32_t XS[8] = { 1, 2, 3, 4, 5, 6, 7, 8 };
uint32_t XD[8] = { 0, 0, 0, 0, 0, 0, 0, 0 };

void test_avx()
{
    __asm__ __volatile__(
        "vmovups 0x00 + %[xs], %%ymm0\n\t"
        "vmovups %%ymm0, 0x00 + %[xd]\n\t"
        :
        : [xs] "o" (XS),
          [xd] "o" (XD)
        : "%xmm0"
    );
}
```

The result:
```
[sadko@freebsd-s ~/tmp]$ clang -O2 -fPIC -o test_sse.clang.o -msse -c test_sse.cpp 
test_sse.cpp:9:39: error: expected relocatable expression
        "movups 0x00 + %[xs], %%xmm0\n\t"
                                      ^
<inline asm>:2:2: note: instantiated into assembly here
        movups %xmm0, 0x00 + (%rax)
        ^
1 error generated.
[sadko@freebsd-s ~/tmp]$ clang -O2 -fPIC -o test_avx.clang.o -mavx -c test_avx.cpp 
test_avx.cpp:9:40: error: expected relocatable expression
        "vmovups 0x00 + %[xs], %%ymm0\n\t"
                                       ^
<inline asm>:2:2: note: instantiated into assembly here
        vmovups %ymm0, 0x00 + (%rax)
        ^
1 error generated.
```

I think I should try to swap the constant and parameter in addition. Maybe this can help

**Comment by sadko4u** [comment]
Probably I can implement some kind of workaround to get rid of offsetable parameters. But this will require some time.

**Comment by yurivict** [comment]
Here is the bug report: https://bugs.llvm.org/show_bug.cgi?id=48197

**Comment by yurivict** [comment]
> Probably I can implement some kind of workaround to get rid of offsetable parameters. But this will require some time.

This is up to you. In the FreeBSD port ```USE_GCC=yes``` workaround is added so that the audio/lsp-plugins port still builds.

**Comment by sadko4u** [comment]
I'm currently working on getting rid of "o" assembly parameters. This may help the code to get compiled with clang.

**Comment by sadko4u** [comment]
Today I've updated lsp-dsp-lib repository which will be part 1.2.0 and now it compiles well with clang.
I'll backport changes to **github-issue-154** branch within few days.

**Comment by sadko4u** [comment]
Pushed changes to **github-issue-154** branch.
Now the problem code compiles well.

Now we have a linker issue:
```
  c++ gen_resources.exe
  c++ lv2_genttl.exe
  c++ vst_genmake.exe
ld: error: unable to find library -liconv
c++: error: linker command failed with exit code 1 (use -v to see invocation)
gmake[2]: *** [Makefile:65: /usr/home/sadko/git/lsp-plugins/.build/vst_genmake.exe] Error 1
gmake[1]: *** [Makefile:71: utils] Error 2
gmake: *** [Makefile:175: compile] Error 2

```

Don't know what clang needs but GCC compiles well and sees the **libiconv** library.

**Comment by yurivict** [comment]
> ld: error: unable to find library -liconv

This is because of the missing -L${PREFIX}/lib. PREFIX=/usr/local

**Comment by sadko4u** [comment]
Tuned makefiles a bit:
```
Build OK
```

Please verify

**Comment by yurivict** [comment]
The github-issue-154 branch builds with clang-10 fine, thanks!

**Comment by sadko4u** [comment]
Merged changes into **devel** branch: 01d628b80b996d9203d9cfa6b97a2773669a45f1

**Comment by sadko4u** [comment]
Removed **github-issue-154** branch.

**Comment by sadko4u** [comment]
Available in 1.1.28 release!
